### PR TITLE
feat: column specs for tabular transformation

### PR DESCRIPTION
**Pull request by sirtorry**
- adds column_specs as an alternative to column_transformation
- adds training_jobs.AutoMLTabularTrainingJob.get_auto_column_specs
- adds aiplatform.column

### CODE REVIEW

File: google/cloud/aiplatform/datasets/tabular_dataset.py
Comment (ivanmkc) [inline google/cloud/aiplatform/datasets/tabular_dataset.py:42]: Start function with a verb?

Comment (sirtorry) [inline google/cloud/aiplatform/datasets/tabular_dataset.py:42]: move auto_column_specs to training_jobs class

Comment (ivanmkc) [reply google/cloud/aiplatform/datasets/tabular_dataset.py:42]: We decided offline to move to the training job and just use the target_column property.

File: google/cloud/aiplatform/datasets/time_series_dataset.py
Comment (sirtorry) [inline google/cloud/aiplatform/datasets/time_series_dataset.py:29]: create a bug, resolve in a different PR.

File: google/cloud/aiplatform/training_jobs.py
Comment (ivanmkc) [inline google/cloud/aiplatform/training_jobs.py:2748]: Class isn’t used so this could be a static method.

File: tests/unit/aiplatform/test_automl_tabular_training_jobs.py
Comment (ivanmkc) [inline tests/unit/aiplatform/test_automl_tabular_training_jobs.py:245]: Why do you set it to numeric and then change it to auto next line?

File: google/cloud/aiplatform/training_jobs.py
Comment (ivanmkc) [inline google/cloud/aiplatform/training_jobs.py:2658]: Could simplify to:
`elif self._column_specs is not None`

File: google/cloud/aiplatform/column.py
Comment (ivanmkc) [inline google/cloud/aiplatform/column.py:21]: I actually do wonder why we don't use an enum instead for this and the schema.

@sasha-gitg did you have a reason why?

**Review by ivanmkc — CHANGES_REQUESTED** [review]
Left comments

File: google/cloud/aiplatform/training_jobs.py
Comment (sirtorry) [reply google/cloud/aiplatform/training_jobs.py:2364]: @sasha-gitg thoughts

File: tests/unit/aiplatform/test_automl_tabular_training_jobs.py
Comment (sirtorry) [reply tests/unit/aiplatform/test_automl_tabular_training_jobs.py:246]: see new assert statement

File: google/cloud/aiplatform/column.py
Comment (sasha-gitg) [reply google/cloud/aiplatform/column.py:21]: This class is used to generate Pipeline Components. AFAIK Pipeline Components do not support Python enums or Python enums nested in a dictionary.

File: google/cloud/aiplatform/training_jobs.py
Comment (sasha-gitg) [inline google/cloud/aiplatform/training_jobs.py:2901]: It's more idiomatic to use `self._column_specs.items()`

File: tests/unit/aiplatform/test_automl_tabular_training_jobs.py
Comment (sasha-gitg) [inline tests/unit/aiplatform/test_automl_tabular_training_jobs.py:351]: We should keep the old tests with column_transformations and create new tests with column specs.

Additionally there should be tests to check the argument validation logic.

File: google/cloud/aiplatform/training_jobs.py
Comment (sasha-gitg) [reply google/cloud/aiplatform/training_jobs.py:2583]: Please also add an example payload of this argument and verbiage on it's format. It's also a bit unintuitive that column_specs and column_transformations have identical docstrings.

File: google/cloud/aiplatform/training_jobs.py
Comment (sasha-gitg) [inline google/cloud/aiplatform/training_jobs.py:2601]: This line should also be included with `column_specs`.

File: google/cloud/aiplatform/__init__.py
Comment (sasha-gitg) [inline google/cloud/aiplatform/__init__.py:95]: This name needs to be qualified a bit more. `aiplatform.data_types` seems too generic. Perhaps, `automl_tabular_column_types`.

File: google/cloud/aiplatform/training_jobs.py
Comment (sasha-gitg) [reply google/cloud/aiplatform/training_jobs.py:2583]: Please provide a full usage example above in the constructor docstring.

**Review by sasha-gitg — CHANGES_REQUESTED** [review]

File: google/cloud/aiplatform/training_jobs.py
Comment (sirtorry) [reply google/cloud/aiplatform/training_jobs.py:2977]: this method is not used during dataset creation, only training job creation. therefore, @ivanmkc believes it should be here.

File: google/cloud/aiplatform/training_jobs.py
Comment (sirtorry) [reply google/cloud/aiplatform/training_jobs.py]: we'll support both for now. we can discuss deprecation when there's a major version bump

File: google/cloud/aiplatform/column.py
Comment (sirtorry) [reply google/cloud/aiplatform/column.py:21]: todo: track with a bug

File: google/cloud/aiplatform/training_jobs.py
Comment (ivanmkc) [reply google/cloud/aiplatform/training_jobs.py:2977]: Yes, we talked about having it on TabularDataset, however, the `column_specs` can only be generated once the `target_column` is known, which is at the training step.

File: google/cloud/aiplatform/training_jobs.py
Comment (ivanmkc) [reply google/cloud/aiplatform/training_jobs.py:2894]: I think that's a fair point, since the service might change in the future.

File: google/cloud/aiplatform/training_jobs.py
Comment (sasha-gitg) [inline google/cloud/aiplatform/training_jobs.py:2588]: It seems confusing to have the same docstring for for both `column_specs` and `column_transformations`. The docstring for `column_specs` should be updated to describe the expected format for `column_spec` along with a reference to accepted transformation values.

File: google/cloud/aiplatform/training_jobs.py
Comment (sasha-gitg) [inline google/cloud/aiplatform/training_jobs.py:2545]: Extending this argument like so, can give users a quick idea of how to format column_specs:
 ```suggestion
            column_specs={ "column_1": "auto", "column_2": "auto"},
```

File: google/cloud/aiplatform/training_jobs.py
Comment (sasha-gitg) [inline google/cloud/aiplatform/training_jobs.py:2891]: It seems like "a" is correct indefinite article unless the guidance from docs is different.

File: google/cloud/aiplatform/training_jobs.py
Comment (sasha-gitg) [inline google/cloud/aiplatform/training_jobs.py:2999]: Where is this being raised?

**Review by sasha-gitg — CHANGES_REQUESTED** [review]

File: google/cloud/aiplatform/training_jobs.py
Comment (ivanmkc) [inline google/cloud/aiplatform/training_jobs.py:2602]: You also might want to note that only columns with no child columns of their own should have a transformation applied.

For example, if you have a column "field_a.field_b", then pass in the transformation for "field_a.field_b" but do not pass in the transformation for "field_a".

File: google/cloud/aiplatform/training_jobs.py
Comment (ivanmkc) [reply google/cloud/aiplatform/training_jobs.py:2891]: Really? Sounds strange to me.

File: tests/unit/aiplatform/test_automl_tabular_training_jobs.py
Comment (ivanmkc) [inline tests/unit/aiplatform/test_automl_tabular_training_jobs.py:595]: Consider adding tests for error cases, such as when both `column_specs` and `column_transformations` are passed. And if `target_column` is not included.

File: google/cloud/aiplatform/training_jobs.py
Comment (ivanmkc) [inline google/cloud/aiplatform/training_jobs.py:3002]: I believe the general thinking is that we defer to the backend to throw these errors instead of checking on the client. You can remove this check.

File: tests/unit/aiplatform/test_automl_tabular_training_jobs.py
Comment (ivanmkc) [inline tests/unit/aiplatform/test_automl_tabular_training_jobs.py:759]: Consider adding a test to provide coverage for the error cases, such as when both `column_specs` and `column_transformations` are provided.

File: google/cloud/aiplatform/training_jobs.py
Comment (ivanmkc) [inline google/cloud/aiplatform/training_jobs.py:2664]: Why not raise Error instead?

Comment (ivanmkc) [reply google/cloud/aiplatform/training_jobs.py:2664]: I think that's more common, see what happens when gcs_source and bq_source are simultaneously passed to a Dataset.

File: google/cloud/aiplatform/training_jobs.py
Comment (sasha-gitg) [inline google/cloud/aiplatform/training_jobs.py:2902]: To simplify, I think we can avoid the _column_specs instance attribute and capture this in the constructor above like so. Also it's more readable to name the tuple values in the list comprehension:

 ```Python
if column_transformations is not None:
    self._column_transformations = column_transformations
elif column_specs is not None:
    self._column_transformations = [
        {transformation: {"column_name": column_name}}
        for column_name, transformation in column_specs.items()
    ]
else:
    self._column_transformations = None
```

**Review by sasha-gitg — APPROVED** [review]
Mainly LGTM. One more comment it seems I didn't include in the last review. Thanks!

**Review by ivanmkc — APPROVED** [review]
LGTM after Sasha's suggestion

**Comment by ivanmkc** [comment]
What about the deprecation warning for column_transformations?

**Pull request merged**
